File: src/paths.ts
import { homedir } from "node:os";
import { join } from "node:path";

export function cronStateDir(): string {
  return process.env.PI_CRON_STATE_DIR || join(process.env.XDG_STATE_HOME || join(homedir(), ".local", "state"), "pi-cron");
}
export function cronRuntimeDir(): string {
  return process.env.PI_CRON_RUNTIME_DIR || (process.env.XDG_RUNTIME_DIR ? join(process.env.XDG_RUNTIME_DIR, "pi-cron") : join(cronStateDir(), "run"));
}
export function socketPath(): string { return process.env.PI_CRON_SOCKET || join(cronRuntimeDir(), "daemon.sock"); }
export function databasePath(): string { return process.env.PI_CRON_DB || join(cronStateDir(), "jobs.sqlite"); }
export function lockPath(): string { return join(cronRuntimeDir(), "daemon.lock"); }

File: rust/src/paths.rs
use std::path::PathBuf;

pub fn state_dir() -> PathBuf {
    std::env::var_os("PI_CRON_STATE_DIR")
        .map(PathBuf::from)
        .unwrap_or_else(|| {
            std::env::var_os("XDG_STATE_HOME")
                .map(PathBuf::from)
                .unwrap_or_else(|| home().join(".local/state"))
                .join("pi-cron")
        })
}
pub fn runtime_dir() -> PathBuf {
    std::env::var_os("PI_CRON_RUNTIME_DIR")
        .map(PathBuf::from)
        .unwrap_or_else(|| {
            std::env::var_os("XDG_RUNTIME_DIR")
                .map(PathBuf::from)
                .map(|p| p.join("pi-cron"))
                .unwrap_or_else(|| state_dir().join("run"))
        })
}
pub fn database() -> PathBuf {
    std::env::var_os("PI_CRON_DB")
        .map(PathBuf::from)
        .unwrap_or_else(|| state_dir().join("jobs.sqlite"))
}
pub fn socket() -> PathBuf {
    std::env::var_os("PI_CRON_SOCKET")
        .map(PathBuf::from)
        .unwrap_or_else(|| runtime_dir().join("daemon.sock"))
}
pub fn lock() -> PathBuf {
    runtime_dir().join("daemon.lock")
}
fn home() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("/tmp"))
}
